Average DSMR telegrams instead of dropping extra ones

dewi-ny-je opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
This doesn't seem to involve a breaking change, since the content of the telegrams is modified, while still based on actual consumption/production data.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Normally the dsmr integration drops all telegrams which are coming faster than the preset interval (standard 30 seconds, so 29 telegrams out of 30).
This is not acceptable whenever average power since the last reading is used to define turning on/off solar inverters to match consumption to production (soon needed in The NEtherlands to keep the lowest prices, once the government subsidies are stopped).
This change will keep track of the extra telegrams usually dropped and provide the resulting averaged readings instead, so that a correct overview of the power flows is maintained. Setting the interval to 0 provides instant data as before.

Fields which are not related to power (textual ones, issue-related ones, and energy ones) are not affected: the only fields affected are those expressed in kW or in W as unit of measure.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

We are talking about the best approach here for people using the DSMR integration to monitor power consumption.

Right now people see one reading every X seconds, depending on the setting, default 60 seconds. It represents only the instant power, but it doesn't give an idea of the overall power consumption: maybe the microwave oven was on during the reading, and 5-10 seconds later the water is done. It looks like high power consumption for 60 seconds, but it was only a moment.
Or a kitchen oven, on most of the time and off maybe 10-20 seconds at time. If you catch the off window, you get a misleading power estimate.
The proposed approach of averaging loses the view of the instant power, but gives a realistic representation of the actual power requirements during the interval set by the user, since peaks are smoothed out.
Moreover, it offers a "graceful degradation", where the readings shown in the interface get progressively less dynamic the longer the interval is set, and yet still representative of the overall power consumption/production of the house. An instantaneous reading, kept for 60 seconds as per default, is really inaccurate.

And not to forget: HA plots graphs as staircase lines: the previous value is maintained **for the whole duration of the measurement interval**. If you plot an instant reading and you maintain it for a period of time, you are literally plotting wrong data.


Disclaimer 1: it is working without crashes on my system and the output graphs are "smoother", indicating the presence of averaging, but I didn't verify manually the averaged data are the correct averages, since this would involve setting up a second USB reader with data splitter and so on. Given the source code I checked, I don't see how it could be wrong, though. Third party review and testing is recommended anyway for such a high profile integration.

Disclaimer 2: this change was not coded by myself (I'm not at that level) but by AI. I asked Gemini 2.5 Pro, 2.5 Flash, Chatgpt o3, o4-mini-high. The proposed modifications were separately provided to gemini 2.5 Pro and chatgpt o3 for comparative review. Both agreed that the code by Gemini 2.5 Pro was the best one, flawless and taking care of edge cases. That is the code I'm currently testing and proposing.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
I have not followed "Create a (feature) branch" because I edited the sensor.py as needed via online interface, so I just followed what Github helped me to do. Otherwise it's a perfect PR.
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
I checked it manually since I'm not practical with ruff, and the differences seem to maintain the style correctly.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

I'll update the docs once the code is approved. It's a small change anyway.


If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

I'm not skilled enough to do that, in everyone's interest.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

dewi-ny-je committed:
Average DSMR datagrams instead of dropping extra ones

Normally the dsmr integration drops all telegrams which are coming faster than the specified interval. This change will keep track of the extra ones and provide averaged readings instead, so that a correct overview of the power consumption is provided. Useful when averaged data is used for defining turning on/off of solar inverters, and other purposes where instant data is not enough.
Setting the interval to 0 provides instant data as before.

Disclaimer 1: it is working without crashes on my system and the output graphs are "smoother", indicating the presence of averaging, but I didn't verify manually the averaged data are the correct averages, since this would involve setting up a second USB reader with data splitter and so on.
Given the source code I checked, I don't see how it could be wrong, though. Third party review and testing is recommended anyway for such a high profile integration.

Disclaimer 2: this change was not coded by myself (I'm not at that level) but by AI. I asked Gemini 2.5 Pro, 2.5 Flash, Chatgpt o3, o4-mini-high. The proposed modifications were separately provided to gemini 2.5 Pro and chatgpt o3 for comparative review. Both agreed that the code by Gemini 2.5 Pro was the best one, flawless and taking care of edge cases. That is the code I'm currently testing and proposing.

elupus commented:
How often do they normally come? Im not sure we need to throttle.

elupus commented:
Ps. This could absolutely make sense to do.

elupus commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -787,25 +805,27 @@ def update_entities_telegram(telegram: Telegram | None) -> None:
         stop_listener = None
         transport = None
-        protocol = None
+        protocol_instance = None  # Renamed from 'protocol'
```
Please revert the rename here. Make it harder to review.

elupus commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -908,7 +928,17 @@ def __init__(
+        self._value: StateType = None  # Stores the value to report
+
+        # Check if this sensor measures power (W or kW)
+        self._is_power_sensor = self.native_unit_of_measurement in {
```
Please generalize here, indicating why the type is specialcased. Like is_averaged_sensor

elupus reviewed (changes requested):
Remove all comments that you changed a line.

dewi-ny-je commented:
> How often do they normally come? Im not sure we need to throttle.

@elupus once per second.

My Xeon won't feel the added load of one telegram per second vs the default of I think 30 seconds, but:
- people might be running much weaker platforms
- one telegram per second on its own adds little overhead but people with automations triggered by them will see a significant increase in computational power requirements 
- the docs explicitly state that "Reducing the default time between updates will increase the amount of events generated and can potentially flood the system with events"
- one telegram per second will make the HA DB grow faster, influxdb will also grow faster
- all of this in exchange for what? there is barely any use case for power readings more often than every 30 seconds (assuming the data returned for power is the correct averaged one and not the instant one), so why should we cause potential issues to so many users, when we could just amend the code to provide meaningful (averaged) power readings?

dewi-ny-je commented:
I'll correct the code in the next few days, thanks for the review

dewi-ny-je committed:
Merge branch 'home-assistant:dev' into patch-1

dewi-ny-je committed:
Correction of raised issues

Reverted "protocol_instance" to "protocol" and renamed "self._is_power_sensor" to "self._is_averaged_sensor"

dewi-ny-je committed:
Merge branch 'patch-1' of https://github.com/dewi-ny-je/core into patch-1

dewi-ny-je commented:
I hope I did it right. I solved the issues raised.
Also, the code has been working in my installation for two-three weeks.

I have no idea how to solve the:

> Check all requirements
> Check hassfest
> Collect information & changes data

dewi-ny-je committed:
Remove unnecessary comments

elupus commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -741,17 +747,24 @@ def init_async_add_entities(telegram: Telegram) -> None:
     )
-    min_time_between_updates = timedelta(
-        seconds=entry.options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
-    )
```
Please revert this move too.

elupus commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -942,54 +972,85 @@ def __init__(
-            return None
+            self._value = None
+        else:
+            self._value = value
```
It feels like we should have a function that "consumes" accumulated values. So it still could be a return value here.

elupus commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -908,7 +928,17 @@ def __init__(
         self._entry = entry
-        self.telegram: Telegram | None = telegram
+        self._telegram: Telegram | None = telegram  # Stores the latest telegram
+        self._obis_reference = entity_description.obis_reference
```
entity description is available in the entity as self.entity_description we don't need to store the obis reference again.

elupus commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -908,7 +928,17 @@ def __init__(
         self._attr_native_unit_of_measurement = native_unit_of_measurement
         self._entry = entry
-        self.telegram: Telegram | None = telegram
+        self._telegram: Telegram | None = telegram  # Stores the latest telegram
```
Still added comment. Also for now avoid renaming with _ prefix so diff is easier to read.

elupus reviewed (changes requested):
We really need tests for this change.

dewi-ny-je committed:
Correct as requested

reverted move of min_time_between_updates
deletion of self._obis_reference

dewi-ny-je committed:
Underscores removed

no underscore in self._telegram

dewi-ny-je committed:
Remove some more comments

dewi-ny-je committed:
Merge branch 'home-assistant:dev' into patch-1

dewi-ny-je committed:
Merge remote-tracking branch 'upstream/dev' into patch-1
